src/dmd/parse.d

```diff
@@ -5851,13 +5935,36 @@ final class Parser(AST) : Lexer
             }
         case TOK.while_:
             {
+                AST.Parameter param = null;
                 nextToken();
                 check(TOK.leftParentheses);
+                param = parseAssignCondition();
                 AST.Expression condition = parseExpression();
                 check(TOK.rightParentheses);
                 Loc endloc;
                 AST.Statement _body = parseStatement(ParseStatementFlags.scope_, null, &endloc);
-                s = new AST.WhileStatement(loc, condition, _body, endloc);
+                /**
+                 * If the while loop is of form `while(auto a = exp) { loop_body }`,
+                 * rewrite to:
+                 *
+                 * while(true)
+                 *     if (auto a = exp)
+                 *     { loop_body }
+                 *     else
+                 *     { break; }
+                 */
+                if (param)
+                {
+                    // end location is from generated code, therefore
+                    // it is set to an initial value so that the
+                    // header generator recognizes this lowering
+                    AST.IfStatement ifd = new AST.IfStatement(loc, param, condition, _body, new AST.BreakStatement(loc, null), Loc());
+                    s = new AST.WhileStatement(loc, new AST.IntegerExp(loc, 1, AST.Type.tbool), ifd, endloc);
```

Comment:
Can't this rewrite wait until the semantic pass? For instance, what if there are semantic errors in the condition (such as unresolved symbols)?

Reply:
I don't see any benefit in waiting up until then. We would need to store `param` for that in the `WhileStatement` and what would that bring us?

To answer your question: the `if` condition is initialized with the start location of the `while` statement, therefore any errors that would occur would be treated the same, except that they are in child scope of the `while` loop. This should not affect in any way error reporting.

Reply:
Please don't lower in the parser.

Reply:
I can do the lowering in the semantic phase, but that will require the addition of a field. That will make things easier for `hdrgen`, but that will make all instances of `WhileStatement` larger.

Reply:
@RazvanN7 - yeah, I was just mindful of the possibility of there being IfStatement-specific error messages that would only serve to confuse the user.

Another example are clients that use the DMD front-end to get symbolic information out of source files (e.g: Editors?).

Reply:
> I can do the lowering in the semantic phase, but that will require the addition of a field. That will make things easier for `hdrgen`, but that will make all instances of `WhileStatement` larger.

I don't think size is a good argument against.  `WhileStatement`'s aren't that important in the bid to keep memory footprint down, unlike `Loc` or `TemplateInstance`.

Reply:
Maybe add the required field to `Statement` it seems it has some space available. And it could serve other purposes too.